be/src/storage/lake/compaction_task_context.h

```diff
@@ -0,0 +1,98 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <butil/containers/linked_list.h>
+#include <rapidjson/document.h>
+#include <rapidjson/rapidjson.h>
+#include <rapidjson/stringbuffer.h>
+#include <rapidjson/writer.h>
+
+#include <memory>
+#include <string>
+
+#include "common/status.h"
+
+namespace starrocks::lake {
+
+class CompactionTaskCallback;
+class Progress {
+public:
+    int value() const { return _value.load(std::memory_order_acquire); }
+
+    void update(int value) { _value.store(value, std::memory_order_release); }
+
+private:
+    std::atomic<int> _value{0};
+};
+
+struct CompactionTaskStats {
+    int64_t io_ns = 0;
+    int64_t segment_init_ns = 0;
+    int64_t column_iterator_init_ns = 0;
+    int64_t io_count_local_disk = 0;
+    int64_t io_count_remote = 0;
+    int64_t compressed_bytes_read = 0;
+    int64_t reader_time_ns = 0;
+    int64_t segment_write_ns = 0;
+};
+
+// Context of a single tablet compaction task.
+struct CompactionTaskContext : public butil::LinkNode<CompactionTaskContext> {
+    explicit CompactionTaskContext(int64_t txn_id_, int64_t tablet_id_, int64_t version_,
+                                   std::shared_ptr<CompactionTaskCallback> cb_)
+            : txn_id(txn_id_), tablet_id(tablet_id_), version(version_), callback(std::move(cb_)) {}
+
+#ifndef NDEBUG
+    ~CompactionTaskContext() {
+        CHECK(next() == this && previous() == this) << "Must remove CompactionTaskContext from list before destructor";
+    }
+#endif
+
+    std::string to_json_stats() {
+        rapidjson::Document root;
+        root.SetObject();
+        auto& allocator = root.GetAllocator();
+        // add stats
+        root.AddMember("reader_total_time_ms", rapidjson::Value(stats->reader_time_ns / 1000000), allocator);
+        root.AddMember("reader_io_ms", rapidjson::Value(stats->segment_init_ns / 1000000), allocator);
+        root.AddMember("reader_io_count_remote", rapidjson::Value(stats->io_count_remote), allocator);
+        root.AddMember("reader_io_count_local_disk", rapidjson::Value(stats->io_count_local_disk), allocator);
+        root.AddMember("compressed_bytes_read", rapidjson::Value(stats->compressed_bytes_read), allocator);
+        root.AddMember("segment_init_ms", rapidjson::Value(stats->segment_init_ns / 1000000), allocator);
+        root.AddMember("column_iterator_init_ms", rapidjson::Value(stats->column_iterator_init_ns / 1000000),
+                       allocator);
+        root.AddMember("segment_write_ms", rapidjson::Value(stats->segment_write_ns / 1000000), allocator);
+
+        rapidjson::StringBuffer strbuf;
+        rapidjson::Writer<rapidjson::StringBuffer> writer(strbuf);
+        root.Accept(writer);
+        return std::string(strbuf.GetString());
```

Comment:
Better move this method definition to cpp file.

Reply:
Code in hpp files is compiled for every cpp file that includes the header. As such, move code to cpp files where possible.